cloudfront-function/main.tf

```diff
@@ -0,0 +1,55 @@
+/**
+ * Config
+ *
+ **/
+
+locals {
+  cloudfront_function_configs = {
+    usage   = var.cloudfront_function_config.usage
+    runtime = var.cloudfront_function_config.runtime
+    code = (
+      var.cloudfront_function_config.usage == "basic_auth" ?
+      file("${path.module}/function.tftpl") :
+      var.cloudfront_function_config.code
+    )
+    basic_auth = {
+      username = try(var.cloudfront_function_config.basic_auth.username, null) != null ? var.cloudfront_function_config.basic_auth.username : var.domain
+      password = try(var.cloudfront_function_config.basic_auth.password, null) != null ? var.cloudfront_function_config.basic_auth.password : random_password.password[0].result
+    }
+  }
+}
+
+resource "aws_cloudfront_function" "function" {
+  publish                      = true
+  name                         = replace(var.domain, ".", "-")
+  comment                      = "function for ${var.domain}"
+  runtime                      = local.cloudfront_function_configs.runtime
+  code                         = local.cloudfront_function_configs.code
+  key_value_store_associations = local.cloudfront_function_configs.usage == "basic_auth" ? [aws_cloudfront_key_value_store.store[0].arn] : []
+}
+
+resource "aws_cloudfront_key_value_store" "store" {
```

Comment:
avoid repeating the resource type in resource names, as it doesn't add any meaningful value. resource **_store - store what exactly? if this is meant to be a global or primary store for the module, the terraform community recommends using names like `main` or `this`. you can find similar naming conventions in popular terraform modules